camel-core/src/main/java/org/apache/camel/processor/CamelInternalProcessor.java

```diff
@@ -854,4 +858,120 @@ public void after(Exchange exchange, Object data) throws Exception {
         }
     }
 
+    /**
+     * Advice for data type contract
+     * TODO add declarative validation
+     */
+    public static class ContractAdvice implements CamelInternalProcessorAdvice {
+        private Contract contract;
+        
+        public ContractAdvice(Contract contract) {
+            this.contract = contract;
+        }
+        
+        @Override
+        public Object before(Exchange exchange) throws Exception {
+            DataType from = getCurrentType(exchange, Exchange.INPUT_TYPE);
+            DataType to = contract.getInputType();
+            if (to != null && !to.equals(from)) {
+                LOG.debug("Looking for transformer for INPUT: from='{}', to='{}'", from, to);
+                convertBody(exchange.getIn(), from, to);
+                exchange.setProperty(Exchange.INPUT_TYPE, to);
+            }
+            return null;
+        }
+        
+        @Override
+        public void after(Exchange exchange, Object data) throws Exception {
+            Message target = exchange.hasOut() ? exchange.getOut() : exchange.getIn();
+            DataType from = getCurrentType(exchange, exchange.hasOut() ? Exchange.OUTPUT_TYPE : Exchange.INPUT_TYPE);
+            DataType to = contract.getOutputType();
+            if (to != null && !to.equals(from)) {
+                LOG.debug("Looking for transformer for OUTPUT: from='{}', to='{}'", from, to);
+                convertBody(target, from, to);
+                exchange.setProperty(exchange.hasOut() ? Exchange.OUTPUT_TYPE : Exchange.INPUT_TYPE, to);
+            }
+        }
+        
+        private static void convertBody(Message message, DataType from, DataType to) throws Exception {
```

Comment:
Can we split this method into smaller methods where each method does one kind of conversion